library/std/src/net/addr.rs

```diff
@@ -290,7 +290,22 @@ impl SocketAddrV4 {
     /// ```
     #[stable(feature = "rust1", since = "1.0.0")]
     pub fn ip(&self) -> &Ipv4Addr {
-        unsafe { &*(&self.inner.sin_addr as *const c::in_addr as *const Ipv4Addr) }
+        // FIXME: what about other OSes besides POSIX and Win32 ?
+
+        // SAFETY: Casting `&in_addr` to `&Ipv4Addr` is safe because:
+        //
+        //   * `Ipv4Addr` is `struct { _: [u8; 4]; }` with size = 4; alignment = 1.
+        //
+        //   * `in_addr` is `struct { _: u32; .. }` in POSIX-compliant OSes [1] and Win32 API [2][3]
+        //      with size and alignment at least 4.
+        //
+        //   * Both are`#[repr(C)]` struct and have the same size (4).
+        //   * Alignment of `Ipv4Addr` (1) is less restrictive than `in_addr` (4).
```

Comment:
```suggestion
        //   * Both are`#[repr(C)]` struct and have the same size (4 bytes).
        //   * Alignment of `Ipv4Addr` (1 byte) is less restrictive than `in_addr` (4 bytes).
```

Should we add the type to make it clearer?

Reply:
The Reference uses the same notion as this PR.